javascripts/2048.js

```diff
@@ -69,76 +72,279 @@ function emptyCell(openCells, usedCells, cellSpace) {
+    }
+  } else if (direction == "right" || direction == 39) {
+    if (tile1[1][1] < tile2[1][1]) {
+      var temp = tile1;
+      tile1 = tile2;
+      tile1 = temp;
+    }
+    var findTile1 = $("div[data-row=" + tile1[0] + "][data-col=" + tile1[1] + "]");
+    var findTile2 = $("div[data-row=" + tile2[0] + "][data-col=" + tile2[1] + "]");
+    var value = $(findTile1).text();
+    if (value == $(findTile2).text()) {
+      $(findTile1).text(value * 2);
+      $(findTile1).attr("data-value", value * 2);
+      deleteTile(findTile2, openCells, usedCells);
+    }
+
+  }
+}
+
+function deleteTile(tile, openCells, usedCells) {
+  var row = $(tile).attr("data-row");
+  var col = $(tile).attr("data-col");
+  $(tile).remove();
+  emptyCell(openCells, usedCells, [row, col]);
+}
+
+
+// MOVING TILES --------------------------------------------------------------------
 function moveTile(tile, openCells, usedCells, direction) {
 
   switch(direction) {
-      case 38: //up
-        return moveUp(tile, openCells, usedCells, direction);
+      case 38: //up or down
+        return verticalMove(tile, openCells, usedCells, direction);
+        break;
+      case 40: //down
+        return verticalMove(tile, openCells, usedCells, direction);
+        break;
+      case 37: //left
+        return horizontalMove(tile, openCells, usedCells, direction);
         break;
+      case 39: //right
+        return horizontalMove(tile, openCells, usedCells, direction);
+        break;
+  }
+}
+
+function horizontalMove(tile, openCells, usedCells, direction) {
+  for ( i = 0; i < 4; i++) {
+    var openRow = openCells.filter(function(array) { return array[0] == "r" + i });
+    var usedRow = usedCells.filter(function(array) { return array[0] == "r" + i });
+    if (usedRow.length >= 4) {
+      //loop through all tiles in row i
+      for (l = 0; l < usedRow.length; l++) {
+        //if there is a next tile after tile l see if you can merge.
+        if (usedRow[l+1] != undefined) {
+          merge(usedRow[l], usedRow[l+1], direction, openCells, usedCells);
+        }
+      }
+    }
+    if (openRow.length >= 4) {
+      continue;
+    } else {
+      if (direction == 37) { //left
+        var openArr = sortArrayByColumn(openRow);
+        var usedArr = sortArrayByColumn(usedRow);
+        for (j = 0; j < usedArr.length; j++) {
+          if (openArr[0][1] < usedArr[j][1]) {
+            var row = "r" + i;
+            oneTile = $("div[data-row=" + row + "][data-col=" + usedArr[j][1] + "]");
+            $(oneTile).attr("data-col", openArr[0][1]);
+            emptyCell(openCells, usedCells, usedArr[j]);
+            emptyCell(openArr, usedArr, usedArr[j]);
+            // Arr before:");
+            // rr);
+            occupyCell(openCells, usedCells, [row, openArr[0][1]]);
+            occupyCell(openArr, usedArr, [openArr[0][0], openArr[0][1]]);
+            // Arr after:");
+            // rr);
+          }
+          usedArr = sortArrayByColumn(usedArr);
```

Comment:
This could be a cause of some of your bugs. You want to avoid mutating your loop control inside the loop itself (you're looping based on `usedArr.length`, but potentially changing the value here)